src/DataStreams/RemoteQueryExecutor.cpp

```diff
@@ -151,6 +151,17 @@ void RemoteQueryExecutor::sendQuery()
     if (settings.skip_unavailable_shards && 0 == multiplexed_connections->size())
         return;
 
+    /// Query cannot be canceled in the middle of the send query,
+    /// since there are multiple packages:
+    /// - Query
+    /// - Data (multiple times)
+    ///
+    /// And after the Cancel packet none Data packet can be sent, otherwise the remote side will throw:
+    ///
+    ///     Unexpected packet Data received from client
+    ///
+    std::lock_guard guard(was_cancelled_mutex);
```

Comment:
It is ok, but not clear why `cancel_mutex` from `MultiplexedConnections` does not help:

https://github.com/ClickHouse/ClickHouse/blob/a06be511df8ebd0a963470026d8460257b891dc7/src/Client/MultiplexedConnections.cpp#L93-L101

Probably, because we send empty block here:
https://github.com/ClickHouse/ClickHouse/blob/b838214a35a84117333e8e5a2764e28e4ec5b952/src/DataStreams/RemoteQueryExecutor.cpp#L167

I suppose it may be more logical to add `sendScalars()` and `sendExternalTables` logic into `MultiplexedConnections::sendQuery`.

Reply:
>Probably, because we send empty block here:

Indeed.

>I suppose it may be more logical to add sendScalars() and sendExternalTables logic into MultiplexedConnections::sendQuery.

Agree